CMakeLists.txt

```diff
@@ -1,74 +1,155 @@
-cmake_minimum_required(VERSION 3.1)
+# Detects whether this is a top-level project
+get_directory_property(LIBIGL_PARENT_DIR PARENT_DIRECTORY)
+if(NOT LIBIGL_PARENT_DIR)
+    set(LIBIGL_TOPLEVEL_PROJECT ON)
+else()
+    set(LIBIGL_TOPLEVEL_PROJECT OFF)
+endif()
+
+# Check required CMake version
+set(REQUIRED_CMAKE_VERSION "3.16.0")
+if(LIBIGL_TOPLEVEL_PROJECT)
+    cmake_minimum_required(VERSION ${REQUIRED_CMAKE_VERSION})
+else()
+    # Don't use cmake_minimum_required here to avoid implicitly overriding parent policies
+    if(${CMAKE_VERSION} VERSION_LESS ${REQUIRED_CMAKE_VERSION})
+        message(FATAL_ERROR "CMake required version to build Libigl is ${REQUIRED_CMAKE_VERSION}")
+    endif()
+endif()
 
-# Toggles the use of the hunter package manager
+# Include user-provided default options if available. We do that before the main
+# `project()` so that we can define the C/C++ compilers from the option file.
+if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/LibiglOptions.cmake)
+    message(STATUS "Using local options file: ${CMAKE_CURRENT_SOURCE_DIR}/LibiglOptions.cmake")
+    include(${CMAKE_CURRENT_SOURCE_DIR}/LibiglOptions.cmake)
+endif()
+
+# If this option is enabled, this will setup the Hunter package manager.
 option(HUNTER_ENABLED "Enable Hunter package manager support" OFF)
+if(HUNTER_ENABLED)
+    include("cmake/misc/HunterGate.cmake")
+    HunterGate(
+        URL "https://github.com/cpp-pm/hunter/archive/v0.23.300.tar.gz"
+        SHA1 "1151d539465d9cdbc880ee30f794864aec11c448"
+    )
+endif()
 
-include("cmake/HunterGate.cmake")
-HunterGate(
-	URL "https://github.com/ruslo/hunter/archive/v0.23.171.tar.gz"
-	SHA1 "5d68bcca78eee347239ca5f4d34f4b6c12683154"
-)
+################################################################################
+project(libigl VERSION 3.0.0)
 
-project(libigl)
+# CMake module path
+list(PREPEND CMAKE_MODULE_PATH
+    ${CMAKE_CURRENT_LIST_DIR}/cmake/
+    ${CMAKE_CURRENT_LIST_DIR}/cmake/igl
+    ${CMAKE_CURRENT_LIST_DIR}/cmake/find
+    ${CMAKE_CURRENT_LIST_DIR}/cmake/recipes/external
+)
 
-# Detects whether this is a top-level project
-get_directory_property(LIBIGL_PARENT_DIR PARENT_DIRECTORY)
-if(NOT LIBIGL_PARENT_DIR)
-	set(LIBIGL_TOPLEVEL_PROJECT ON)
+if(HUNTER_ENABLED)
+    list(PREPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake/recipes/hunter)
 else()
-	set(LIBIGL_TOPLEVEL_PROJECT OFF)
+    list(PREPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake/recipes/external)
+endif()
+
+# Check for gmp + mpfr and enable cork / cgal modules accordingly
+set(LIBIGL_DEFAULT_CGAL_CORK ${LIBIGL_TOPLEVEL_PROJECT})
```

Comment:
We don't have an option `LIBIGL_DEFAULT_GMP_MPFR`. Here `LIBIGL_DEFAULT_CGAL_CORK` controls the default value of the options for the CGAL and Cork modules. If GMP and MPFR are not available on the target machine, then I disable the CGAL/Cork modules by default, since CGAL/Cork depend on GMP+MPFR.

Reply:
Right, and I'm commenting that the name `LIBIGL_DEFAULT_CGAL_CORK` is confusing since it seems to primarily refer to the presence of gmp/mpfr.

Reply:
The value of `LIBIGL_DEFAULT_CGAL_CORK` is set based on the presence of gmp/mpfr, but what it is used for is to set the default option for the CGAL and Cork modules. But hey I'm happy to rename it if you prefer.

Reply:
One thing though: if libigl is not built as a subproject, `LIBIGL_DEFAULT_CGAL_CORK` will be OFF. Also on Windows, `LIBIGL_DEFAULT_CGAL_CORK` will always be ON by default for toplevel project. So if I rename the variable into `LIBIGL_GMP_MPFR_AVAILABLE`, this would be somewhat misleading (e.g. GMP/MPFR might be available, but because libigl is built as a subproject, it will be set to OFF).

Reply:
I see. Maybe it makes sense to have a variable that tracks whether GMP_MPFR has been found and then the existing variables for wether building the cork / cgal modules.

Reply:
> a variable that tracks whether GMP_MPFR has been found

What would you use it for? To track whether GMP/MPFR have been found, I would test for the existence of the imported targets `gmp::gmp` and `mpfr::mpfr`. Using variables in CMake is kinda brittle so I like to keep it to the bare minimum. But if you want to amend this part of the script to your liking feel free to give it a shot, since I'm not sure what you have in mind.

Reply:
I don't think I can amend it because I don't understand it. What does `LIBIGL_DEFAULT_CGAL_CORK`  mean? `LIBIGL` I can guess means this is one of our variables, `DEFAULT` means this is a flag On or Off for a default setting. `CGAL_CORK` why both? Is this variable the same as `LIBIGL_DEFAULT_CGAL AND LIBIGL_DEFAULT_CORK`?

Reply:
I mean it's just a variable name. The "correct" name should be:
```
LIBIGL_DEFAULT_VALUE_FOR_CMAKE_OPTION_CGAL_AND_CORK_MODULES_BASED_ON_WHETHER_GMP_AND_MPFR_ARE_AVAILABLE
```

Then I just use this variable twice to set the default value for the following CMake options:
- LIBIGL_COPYLEFT_WITH_CGAL
- LIBIGL_COPYLEFT_WITH_CORK

So really this should be invisible to the user. LIBIGL_DEFAULT_CGAL_CORK is just a "local variable", it's not a proper CMake option that the user has to concern themselves with. If the user provides their own value for LIBIGL_COPYLEFT_WITH_CGAL or LIBIGL_COPYLEFT_WITH_CORK, the user-provided option will take precedence.